packages/openapi-v3/src/decorators/request-body.spike.decorator.ts

```diff
@@ -0,0 +1,124 @@
+  [TS_TYPE_KEY]?: Function,
+  isVisited?: boolean,
+}
+
+function isRequestBodySpec(spec: Partial<RequestBodyObject> | SchemaOptions): spec is Partial<RequestBodyObject> {
+  return (spec as Partial<RequestBodyObject>).description !== undefined
+    || (spec as Partial<RequestBodyObject>).required !== undefined
+    || (spec as Partial<RequestBodyObject>).content !== undefined;
+}
+
+export function requestBody2(
+  specOrModelOrOptions?: Partial<RequestBodyObject> | Function | SchemaOptions,
+  modelOrOptions?: Function | SchemaOptions,
+  schemaOptions?: SchemaOptions
+) {
+  if (typeof specOrModelOrOptions == 'function') {
+    // omit the 1st param
+    // @requestBody(modelCtor, schemaOptions)
+    if (modelOrOptions && typeof modelOrOptions == 'object')
+      return _requestBody2({}, specOrModelOrOptions, modelOrOptions as SchemaOptions);
+    // @requestBody(modelCtor)
+    return _requestBody2({}, specOrModelOrOptions);
+  } else if (specOrModelOrOptions && isRequestBodySpec(specOrModelOrOptions)) {
+    // 1st param is spec
+    if (modelOrOptions && typeof modelOrOptions == 'object')
+      // omit the 2nd param
+      // @requestBody(spec, schemaOptions)
+      return _requestBody2(specOrModelOrOptions, undefined, modelOrOptions)
+    // @requestBody(spec)
+    // @requestBody(spec, modelCtor)
+    // @requestBody(spec, modelCtor, schemaOptions)
+    return _requestBody2(specOrModelOrOptions, modelOrOptions, schemaOptions);
+  } else if (specOrModelOrOptions !== undefined) {
+    // omit 1st and 2nd params
+    // @requestBody(schemaOptions)
+    return _requestBody2({}, undefined, specOrModelOrOptions as SchemaOptions);
+  } else {
+    // no params
+    // @requestBody()
+    return _requestBody2();
+  }
+}
+
+// `name` is provided to avoid generating the same schema
+function _requestBody2(
+  requestBodySpecification?: Partial<RequestBodyObject>,
+  modelCtor?: Function,
+  schemaOptions?: SchemaOptions
+) {
+  return function (target: object, member: string, index: number) {
+    debug('@newRequestBody() on %s.%s', target.constructor.name, member);
+    debug('  parameter index: %s', index);
+    /* istanbul ignore if */
+    if (debug.enabled)
+      debug('  schemaOptions: %s', inspect(requestBodySpecification, { depth: null }));
+
+    // Use 'application/json' as default content if `requestBody` is undefined
+    const requestBodySpec = Object.assign({}, requestBodySpecification || { content: {} });
+
+    if (_.isEmpty(requestBodySpec.content))
+      requestBodySpec.content = { 'application/json': {} };
+
+    // Get the design time method parameter metadata
+    const methodSig = MetadataInspector.getDesignTypeForMethod(target, member);
+    const paramTypes = (methodSig && methodSig.parameterTypes) || [];
+
+    const paramType = paramTypes[index];
+
+    // preserve backward compatibility
+    if (modelCtor) schemaOptions = Object.assign({}, schemaOptions, { [TS_TYPE_KEY]: modelCtor });
```

Comment:
Let's not use `TS_TYPE_KEY` please and find a way how to leverage `getModelSchemaRef` instead.

Ideally, I'd like all of the changes to stay inside the request-body decorator, we should not need to change other functions like `resolveSchema`.

Would the following work?

```ts
const ctor = modelCtor || paramCtor;
if (!ctor) {
  // No model to infer the schema from. This means the user is supposed
  // to describe content schema in requestBodySpecification.
  // Print a warning or throw an error if they did not do so,
  // because the OpenAPI spec  is not very useful without content schema
  return;
}
 
const schema = getModelSchemaRef(ctor, schemaOptions);
requestBodySpec.content = _.mapValues(requestBodySpec.content, c => {
  if (!c.schema) {
    c.schema = schema;
  }
 return c;
});
```

Reply:
reverted this code as well in https://github.com/strongloop/loopback-next/pull/3466